python/lsst/jointcal/jointcal.py

```diff
@@ -174,6 +174,42 @@ def lookupStaticCalibrations(datasetType, registry, quantumDataId, collections):
                                   findFirst=True)
 
 
+def lookupVisitRefCats(datasetType, registry, quantumDataId, collections):
+    """Lookup function that finds all refcats for all visits that overlap a
+    tract, rather than just the refcats that directly overlap the tract.
```

Comment:
Assuming I understand the logic here, I believe this will work for the DM-28991 bug, but I don't actually understand how the refcat<->region<->visit relationships are defined. I wish we had a way we could test this kind of logic, without constructing a multi-tract dataset.

Reply:
The only ideas I have for regular testing involve a multi-tract dataset, but only the database side of it, not the files - for example, there's a YAML export file committed to daf_butler (in `tests/data/registry`) that's a dump of enough of the real RC2 registry to test a lot of these problems, and without the actual data files it's pretty compact and pretty fast to import into a local SQLite registry.  And `tests/test_simpleButler.py` uses that to good effect with a mocked `Datastore` that doesn't care about whether the files exist.  But it might be a lot harder to extend the mocking all the way through to the Python objects a `PipelineTask` actually expects to get back from the butler, so I'm not sure it's actually worth it.

Another approach would be to test this via the RC2 runs themselves, maybe via some kind of metric-generating `PipelineTask` that inspects the jointcal outputs and checks that the outputs satisfy various conditions.  I don't know if jointcal actually generates any outputs that would let such a task look for which stars got refcat matches, but that'd be a good thing to have anyway.

And I should add that I also don't think that CI testing for this is a super high priority; I'm sure the (one-off) Gen2->Gen3 scientific parity testing we're doing would reveal this if it's _currently_ still a problem, and it's the kind of thing that's unlikely to stop working silently.

I'm also happy to explain more about how the refcat<->region<->visit relationships are defined, but maybe it'd be better for you to post that as a question in `#dm-middleware-support` or community in order to get more eyes on something that's probably of more general interest.

Reply:
jointcal computes a number of metrics that could be used to check this, but they are not output in gen3 yet (see DM-28863). `collected_astrometry_refStars` would be a value to compare directly between gen2 and gen3: those should be identical across middleware if this fix works. Maybe this means that DM-28863 should be a higher priority now, so we can do that comparison?

Reply:
Yeah, DM-28863 does seem like something it might be nice to have in hand sometime in the next couple of months, and certainly something we could use to quickly retire DM-28991 when it's in place.  The next two RC2 runs are w18 and w22, so either would be a good target. I'm hoping we'll learn in the next week or so whether there are big differences between jointcal outputs between Gen2 and Gen3, and the nature of those differences (if there are any) would inform what kinds of metrics we'd want to include in w18 (or run sooner against the w14+DM-29670 outputs).